pkg/alertmanager/validation.go

```diff
@@ -218,6 +218,19 @@ func validateAlertManagerRoutes(r *monitoringv1alpha1.Route, receivers map[strin
 		return errors.Errorf("receiver %q not found", r.Receiver)
 	}
 
+	if groupLen := len(r.GroupBy); groupLen > 0 {
```

Comment:
For my understanding, what is the reason for using both the OpenAPI validation and implementing our own operator-side validation?

Reply:
From my understanding, there are at least a couple of reasons:

1. There are some cases that the OpenAPi validation (at least with the kubebuilder annotations) do not cover. This PR actually contains an example of one of those. We want unique items in the slice (covered with `UniqueItems`), but we also want to validate that if `'...'` is in the list, there are no other items, so we need to validate ourselves server-side (unless I've missed something).
2. We implement a validating webhook, so it made sense to split out the entire validation and have the admission controller call that.

Reply:
The error message from the failing test is as follows: `alertmanager_instance_namespaces_test.go:136: initialize AlertmanagerConfig CRD: create CRD alertmanagerconfigs on the apiserver: create CRD: AlertmanagerConfig: CustomResourceDefinition.apiextensions.k8s.io "alertmanagerconfigs.monitoring.coreos.com" is invalid: spec.validation.openAPIV3Schema.properties[spec].properties[route].properties[groupBy].uniqueItems: Forbidden: uniqueItems cannot be set to true since the runtime complexity becomes quadratic`

I went digging around and found https://github.com/kubernetes/community/pull/708/ and https://github.com/kubernetes/kubernetes/commit/fd09c3dbb67374e4f4b46b820b5cd3d9a3124cda#diff-6d0d1a9c59e6d17eea2b3ccdf2d6eb7d6169efdf7ed5d5e1f5c09ce5632efd18R205 which basically means we cannot have the `UniqueItems:=true`, so again a custom validation func is required. 

I am surprised this is not supported for basic types like int and string but here we are.

Reply:
Alright, this explains it then. Thanks for the input.

Reply:
Make it list-type set and it will be checked for duplicates.